engine/src/main/java/org/terasology/entitySystem/event/AbstractBeforeAfterEvent.java

```diff
@@ -0,0 +1,42 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.entitySystem.event;
+
+/**
+ * Immutable event to notify a change in a certain value.
+ * NOTE: This is a generic event and a more specific Event extending this event should be fired and reacted to.
+ */
+public class AbstractBeforeAfterEvent implements Event {
+    private final float oldValue;
+    private final float newValue;
+
+    /**
+     * Creates a new notification event on change in value.
+     */
+    AbstractBeforeAfterEvent(float oldValue, float newValue) {
```

Comment:
Same here:
```suggestion
    AbstractBeforeAfterEvent(final T oldValue, final T newValue) {
```

We can also mark them final as we are not going to change them ever in this event.